File: src/models/action-rules.ts
/**
 * Action Rules System
 *
 * A unified system for mapping buttons to actions, supporting:
 * - Individual button-to-action mappings with triggers (press/release/hold)
 * - Button groups with group-specific actions (e.g., chord progressions)
 * - Startup rules (button-less actions that run on initialization)
 * - Button naming for user clarity
 */

import { ActionDefinition } from '../core/actions.js';

/**
 * Button identifier type
 * - "button:primary" - Stylus primary button
 * - "button:secondary" - Stylus secondary button
 * - "button:1" through "button:N" - Tablet buttons by number
 */
export type ButtonId = `button:${string}`;

/**
 * Trigger type for when an action should execute
 */
export type TriggerType = 'press' | 'release' | 'hold';

/**
 * Action category - distinguishes between button and group actions
 */
export type ActionCategory = 'button' | 'group' | 'startup';

/**
 * Group action type - actions that can be assigned to button groups
 * Currently only chord-progression is supported
 */
export type GroupActionType = 'chord-progression';

/**
 * Group action definition - action with parameters for button groups
 */
export interface GroupAction {
  /** Type of group action */
  type: GroupActionType;
  /** Chord progression preset name (for chord-progression type) */
  progression: string;
  /** Octave for chord playback (for chord-progression type) */
  octave: number;
}

/**
 * Individual action rule - maps a button to an action
 */
export interface ActionRule {
  /** Unique identifier for this rule */
  id: string;
  /** Optional human-readable name for this rule */
  name?: string;
  /** Button that triggers this action */
  button: ButtonId;
  /** Action to execute (string or array with params) */
  action: ActionDefinition;
  /** When to trigger: 'press', 'release' (default), or 'hold' */
  trigger?: TriggerType;
}

/**
 * Button group - a named collection of buttons
 * Actions are assigned to groups via GroupRule, not directly on the group
 */
export interface ButtonGroup {
  /** Unique identifier for this group */
  id: string;
  /** Human-readable name for this group */
  name: string;
  /** Buttons in this group (in order for action mapping) */
  buttons: ButtonId[];
}

/**
 * Group rule - assigns a group action to a button group
 */
export interface GroupRule {
  /** Unique identifier for this rule */
  id: string;
  /** Optional human-readable name for this rule */
  name?: string;
  /** Group that this rule applies to */
  groupId: string;
  /** Group action to execute */
  action: GroupAction;
}

/**
 * Startup rule - an action that executes on initialization (no button)
 */
export interface StartupRule {
  /** Unique identifier for this rule */
  id: string;
  /** Human-readable name for this rule */
  name: string;
  /** Action to execute */
  action: ActionDefinition;
}

/**
 * Complete action rules configuration
 */
export interface ActionRulesConfigData {
  /** Human-readable names for buttons */
  buttonNames: Record<ButtonId, string>;
  /** Individual button-to-action rules */
  rules: ActionRule[];
  /** Button groups (just collections of buttons) */
  groups: ButtonGroup[];
  /** Group rules - actions assigned to groups */
  groupRules: GroupRule[];
  /** Actions that execute on startup */
  startupRules: StartupRule[];
}

/**
 * Default action rules configuration
 */
export const DEFAULT_ACTION_RULES_CONFIG: ActionRulesConfigData = {
  buttonNames: {},
  rules: [],
  groups: [],
  groupRules: [],
  startupRules: [],
};

/**
 * Generate a unique ID for rules/groups
 */
export function generateRuleId(prefix: string = 'rule'): string {
  return `${prefix}-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
}

/**
 * Parse a button ID into its components
 */
export function parseButtonId(buttonId: ButtonId): { type: 'stylus' | 'tablet'; identifier: string } {
  const parts = buttonId.split(':');
  if (parts.length !== 2 || parts[0] !== 'button') {
    throw new Error(`Invalid button ID: ${buttonId}`);
  }
  
  const identifier = parts[1];
  if (identifier === 'primary' || identifier === 'secondary') {
    return { type: 'stylus', identifier };
  }
  
  // Numeric tablet button
  const num = parseInt(identifier, 10);
  if (isNaN(num) || num < 1) {
    throw new Error(`Invalid button ID: ${buttonId}`);
  }
  
  return { type: 'tablet', identifier };
}

/**
 * Create a button ID from components
 */
export function createButtonId(type: 'stylus' | 'tablet', identifier: string | number): ButtonId {
  if (type === 'stylus') {
    if (identifier !== 'primary' && identifier !== 'secondary') {
      throw new Error(`Invalid stylus identifier: ${identifier}`);
    }
    return `button:${identifier}` as ButtonId;
  }
  
  return `button:${identifier}` as ButtonId;
}

/**
 * Get a human-readable label for a button ID
 */
export function getButtonLabel(buttonId: ButtonId, buttonNames?: Record<ButtonId, string>): string {
  // Check for custom name first
  if (buttonNames && buttonNames[buttonId]) {
    return buttonNames[buttonId];
  }
  
  // Generate default label
  const { type, identifier } = parseButtonId(buttonId);
  if (type === 'stylus') {
    return identifier === 'primary' ? 'Primary Stylus' : 'Secondary Stylus';
  }
  
  return `Button ${identifier}`;
}

/**
 * Action Rules Configuration class
 */
export class ActionRulesConfig implements ActionRulesConfigData {
  buttonNames: Record<ButtonId, string>;
  rules: ActionRule[];
  groups: ButtonGroup[];
  groupRules: GroupRule[];
  startupRules: StartupRule[];

  constructor(data: Partial<ActionRulesConfigData> = {}) {
    this.buttonNames = { ...DEFAULT_ACTION_RULES_CONFIG.buttonNames, ...data.buttonNames };
    this.rules = data.rules ? [...data.rules] : [];
    this.groups = data.groups ? [...data.groups] : [];
    this.groupRules = data.groupRules ? [...data.groupRules] : [];
    this.startupRules = data.startupRules ? [...data.startupRules] : [];
  }

  /**
   * Create from a plain object (e.g., from JSON)
   */
  static fromDict(data: Record<string, unknown> | null | undefined): ActionRulesConfig {
    if (!data) {
      return new ActionRulesConfig();
    }

    return new ActionRulesConfig({
      buttonNames: (data.buttonNames ?? data.button_names) as Record<ButtonId, string> | undefined,
      rules: (data.rules as ActionRule[]) ?? [],
      groups: (data.groups as ButtonGroup[]) ?? [],
      groupRules: (data.groupRules ?? data.group_rules) as GroupRule[] | undefined,
      startupRules: (data.startupRules ?? data.startup_rules) as StartupRule[] | undefined,
    });
  }

  /**
   * Convert to a plain object for JSON serialization
   */
  toDict(): ActionRulesConfigData {
    return {
      buttonNames: { ...this.buttonNames },
      rules: [...this.rules],
      groups: [...this.groups],
      groupRules: [...this.groupRules],
      startupRules: [...this.startupRules],
    };
  }

  /**
   * Add a new action rule
   */
  addRule(rule: Omit<ActionRule, 'id'> & { id?: string }): ActionRule {
    const newRule: ActionRule = {
      ...rule,
      id: rule.id ?? generateRuleId('rule'),
      trigger: rule.trigger ?? 'release',
    };
    this.rules.push(newRule);
    return newRule;
  }

  /**
   * Remove a rule by ID
   */
  removeRule(ruleId: string): boolean {
    const index = this.rules.findIndex(r => r.id === ruleId);
    if (index >= 0) {
      this.rules.splice(index, 1);
      return true;
    }
    return false;
  }

  /**
   * Update a rule by ID
   */
  updateRule(ruleId: string, updates: Partial<Omit<ActionRule, 'id'>>): boolean {
    const rule = this.rules.find(r => r.id === ruleId);
    if (rule) {
      Object.assign(rule, updates);
      return true;
    }
    return false;
  }

  /**
   * Add a new button group
   */
  addGroup(group: Omit<ButtonGroup, 'id'> & { id?: string }): ButtonGroup {
    const newGroup: ButtonGroup = {
      ...group,
      id: group.id ?? generateRuleId('group'),
    };
    this.groups.push(newGroup);
    return newGroup;
  }

  /**
   * Remove a group by ID
   */
  removeGroup(groupId: string): boolean {
    const index = this.groups.findIndex(g => g.id === groupId);
    if (index >= 0) {
      this.groups.splice(index, 1);
      return true;
    }
    return false;
  }

  /**
   * Update a group by ID
   */
  updateGroup(groupId: string, updates: Partial<Omit<ButtonGroup, 'id'>>): boolean {
    const group = this.groups.find(g => g.id === groupId);
    if (group) {
      Object.assign(group, updates);
      return true;
    }
    return false;
  }

  /**
   * Add a new group rule
   */
  addGroupRule(rule: Omit<GroupRule, 'id'> & { id?: string }): GroupRule {
    const newRule: GroupRule = {
      ...rule,
      id: rule.id ?? generateRuleId('grouprule'),
    };
    this.groupRules.push(newRule);
    return newRule;
  }

  /**
   * Remove a group rule by ID
   */
  removeGroupRule(ruleId: string): boolean {
    const index = this.groupRules.findIndex(r => r.id === ruleId);
    if (index >= 0) {
      this.groupRules.splice(index, 1);
      return true;
    }
    return false;
  }

  /**
   * Update a group rule by ID
   */
  updateGroupRule(ruleId: string, updates: Partial<Omit<GroupRule, 'id'>>): boolean {
    const rule = this.groupRules.find(r => r.id === ruleId);
    if (rule) {
      Object.assign(rule, updates);
      return true;
    }
    return false;
  }

  /**
   * Get the group rule for a specific group (if any)
   */
  getGroupRuleForGroup(groupId: string): GroupRule | undefined {
    return this.groupRules.find(r => r.groupId === groupId);
  }

  /**
   * Add a startup rule
   */
  addStartupRule(rule: Omit<StartupRule, 'id'> & { id?: string }): StartupRule {
    const newRule: StartupRule = {
      ...rule,
      id: rule.id ?? generateRuleId('startup'),
    };
    this.startupRules.push(newRule);
    return newRule;
  }

  /**
   * Remove a startup rule by ID
   */
  removeStartupRule(ruleId: string): boolean {
    const index = this.startupRules.findIndex(r => r.id === ruleId);
    if (index >= 0) {
      this.startupRules.splice(index, 1);
      return true;
    }
    return false;
  }

  /**
   * Set a button's name
   */
  setButtonName(buttonId: ButtonId, name: string): void {
    if (name) {
      this.buttonNames[buttonId] = name;
    } else {
      delete this.buttonNames[buttonId];
    }
  }

  /**
   * Get all rules for a specific button
   */
  getRulesForButton(buttonId: ButtonId): ActionRule[] {
    return this.rules.filter(r => r.button === buttonId);
  }

  /**
   * Get the group that contains a specific button (if any)
   */
  getGroupForButton(buttonId: ButtonId): ButtonGroup | undefined {
    return this.groups.find(g => g.buttons.includes(buttonId));
  }

  /**
   * Get the action for a button press/release/hold event
   * Returns the action definition if found, or null if no matching rule
   */
  getActionForButtonEvent(buttonId: ButtonId, trigger: TriggerType): ActionDefinition | null {
    // First check individual rules
    const rule = this.rules.find(r => r.button === buttonId && (r.trigger ?? 'release') === trigger);
    if (rule) {
      return rule.action;
    }

    // Then check groups (groups only respond to 'release' by default)
    if (trigger === 'release') {
      const group = this.getGroupForButton(buttonId);
      if (group) {
        const buttonIndex = group.buttons.indexOf(buttonId);
        if (buttonIndex >= 0) {
          // Find the group rule for this group
          const groupRule = this.getGroupRuleForGroup(group.id);
          if (groupRule) {
            // Handle group action based on type
            if (groupRule.action.type === 'chord-progression') {
              // Return a set-chord-in-progression action
              return ['set-chord-in-progression', groupRule.action.progression, buttonIndex, groupRule.action.octave];
            }
          }
        }
      }
    }

    return null;
  }
}

File: python/sketchatone/models/action_rules.py
"""
Action Rules System

A unified system for mapping buttons to actions, supporting:
- Individual button-to-action mappings with triggers (press/release/hold)
- Button groups with group-specific actions (e.g., chord progressions)
- Startup rules (button-less actions that run on initialization)
- Button naming for user clarity

Ported from TypeScript: src/models/action-rules.ts
"""

from dataclasses import dataclass, field
from typing import Dict, Any, List, Optional, Literal, Union
import time
import random
import string


# Type aliases
ButtonId = str  # Format: "button:primary", "button:secondary", "button:1", etc.
TriggerType = Literal['press', 'release', 'hold']
ActionCategory = Literal['button', 'group', 'startup']
GroupActionType = Literal['chord-progression']

# Action definition can be a string or list with params
ActionDefinition = Union[str, List[Any], None]


@dataclass
class GroupAction:
    """
    Group action definition - action with parameters for button groups.
    Currently only chord-progression is supported.
    """
    type: GroupActionType
    progression: str  # Chord progression preset name
    octave: int = 4   # Octave for chord playback

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'GroupAction':
        """Create from dictionary"""
        return cls(
            type=data.get('type', 'chord-progression'),
            progression=data.get('progression', 'c-major-pop'),
            octave=data.get('octave', 4)
        )

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for JSON serialization"""
        return {
            'type': self.type,
            'progression': self.progression,
            'octave': self.octave
        }


@dataclass
class ActionRule:
    """
    Individual action rule - maps a button to an action.
    """
    id: str
    button: ButtonId
    action: ActionDefinition
    name: Optional[str] = None
    trigger: TriggerType = 'release'

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'ActionRule':
        """Create from dictionary (supports both snake_case and camelCase)"""
        return cls(
            id=data.get('id', generate_rule_id('rule')),
            button=data.get('button', ''),
            action=data.get('action'),
            name=data.get('name'),
            trigger=data.get('trigger', 'release')
        )

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for JSON serialization"""
        result = {
            'id': self.id,
            'button': self.button,
            'action': self.action,
            'trigger': self.trigger
        }
        if self.name:
            result['name'] = self.name
        return result


@dataclass
class ButtonGroup:
    """
    Button group - a named collection of buttons.
    Actions are assigned to groups via GroupRule, not directly on the group.
    """
    id: str
    name: str
    buttons: List[ButtonId] = field(default_factory=list)

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'ButtonGroup':
        """Create from dictionary"""
        return cls(
            id=data.get('id', generate_rule_id('group')),
            name=data.get('name', ''),
            buttons=data.get('buttons', [])
        )

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for JSON serialization"""
        return {
            'id': self.id,
            'name': self.name,
            'buttons': self.buttons
        }


@dataclass
class GroupRule:
    """
    Group rule - assigns a group action to a button group.
    """
    id: str
    group_id: str
    action: GroupAction
    name: Optional[str] = None
    trigger: TriggerType = 'release'

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'GroupRule':
        """Create from dictionary (supports both snake_case and camelCase)"""
        action_data = data.get('action', {})
        return cls(
            id=data.get('id', generate_rule_id('grouprule')),
            group_id=data.get('group_id', data.get('groupId', '')),
            action=GroupAction.from_dict(action_data) if isinstance(action_data, dict) else GroupAction(type='chord-progression', progression='c-major-pop'),
            name=data.get('name'),
            trigger=data.get('trigger', 'release')
        )

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for JSON serialization (camelCase for webapp)"""
        result = {
            'id': self.id,
            'groupId': self.group_id,
            'action': self.action.to_dict(),
            'trigger': self.trigger
        }
        if self.name:
            result['name'] = self.name
        return result


@dataclass
class StartupRule:
    """
    Startup rule - an action that executes on initialization (no button).
    """
    id: str
    name: str
    action: ActionDefinition

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'StartupRule':
        """Create from dictionary"""
        return cls(
            id=data.get('id', generate_rule_id('startup')),
            name=data.get('name', ''),
            action=data.get('action')
        )

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for JSON serialization"""
        return {
            'id': self.id,
            'name': self.name,
            'action': self.action
        }


def generate_rule_id(prefix: str = 'rule') -> str:
    """Generate a unique ID for rules/groups"""
    timestamp = int(time.time() * 1000)
    random_str = ''.join(random.choices(string.ascii_lowercase + string.digits, k=9))
    return f"{prefix}-{timestamp}-{random_str}"


def parse_button_id(button_id: ButtonId) -> Dict[str, str]:
    """
    Parse a button ID into its components.
    
    Returns:
        Dict with 'type' ('stylus' or 'tablet') and 'identifier'
    """
    parts = button_id.split(':')
    if len(parts) != 2 or parts[0] != 'button':
        raise ValueError(f"Invalid button ID: {button_id}")
    
    identifier = parts[1]
    if identifier in ('primary', 'secondary'):
        return {'type': 'stylus', 'identifier': identifier}
    
    # Numeric tablet button
    try:
        num = int(identifier)
        if num < 1:
            raise ValueError(f"Invalid button ID: {button_id}")
        return {'type': 'tablet', 'identifier': identifier}
    except ValueError:
        raise ValueError(f"Invalid button ID: {button_id}")


def create_button_id(button_type: str, identifier: Union[str, int]) -> ButtonId:
    """Create a button ID from components"""
    if button_type == 'stylus':
        if identifier not in ('primary', 'secondary'):
            raise ValueError(f"Invalid stylus identifier: {identifier}")
    return f"button:{identifier}"


def get_button_label(button_id: ButtonId, button_names: Optional[Dict[ButtonId, str]] = None) -> str:
    """Get a human-readable label for a button ID"""
    # Check for custom name first
    if button_names and button_id in button_names:
        return button_names[button_id]
    
    # Generate default label
    parsed = parse_button_id(button_id)
    if parsed['type'] == 'stylus':
        return 'Primary Stylus' if parsed['identifier'] == 'primary' else 'Secondary Stylus'
    
    return f"Button {parsed['identifier']}"


class ActionRulesConfig:
    """
    Action Rules Configuration class.
    Manages button-to-action mappings, groups, and startup rules.
    """
    
    def __init__(
        self,
        button_names: Optional[Dict[ButtonId, str]] = None,
        rules: Optional[List[ActionRule]] = None,
        groups: Optional[List[ButtonGroup]] = None,
        group_rules: Optional[List[GroupRule]] = None,
        startup_rules: Optional[List[StartupRule]] = None
    ):
        self.button_names: Dict[ButtonId, str] = button_names or {}
        self.rules: List[ActionRule] = rules or []
        self.groups: List[ButtonGroup] = groups or []
        self.group_rules: List[GroupRule] = group_rules or []
        self.startup_rules: List[StartupRule] = startup_rules or []

    @classmethod
    def from_dict(cls, data: Optional[Dict[str, Any]]) -> 'ActionRulesConfig':
        """Create from a plain object (e.g., from JSON)"""
        if not data:
            return cls()
        
        # Parse button names (supports both snake_case and camelCase)
        button_names = data.get('button_names', data.get('buttonNames', {}))
        
        # Parse rules
        raw_rules = data.get('rules', [])
        rules = [ActionRule.from_dict(r) for r in raw_rules]
        
        # Parse groups
        raw_groups = data.get('groups', [])
        groups = [ButtonGroup.from_dict(g) for g in raw_groups]
        
        # Parse group rules (supports both snake_case and camelCase)
        raw_group_rules = data.get('group_rules', data.get('groupRules', []))
        group_rules = [GroupRule.from_dict(gr) for gr in raw_group_rules]
        
        # Parse startup rules (supports both snake_case and camelCase)
        raw_startup_rules = data.get('startup_rules', data.get('startupRules', []))
        startup_rules = [StartupRule.from_dict(sr) for sr in raw_startup_rules]
        
        return cls(
            button_names=button_names,
            rules=rules,
            groups=groups,
            group_rules=group_rules,
            startup_rules=startup_rules
        )

    def to_dict(self) -> Dict[str, Any]:
        """Convert to a plain object for JSON serialization (camelCase for webapp)"""
        return {
            'buttonNames': self.button_names,
            'rules': [r.to_dict() for r in self.rules],
            'groups': [g.to_dict() for g in self.groups],
            'groupRules': [gr.to_dict() for gr in self.group_rules],
            'startupRules': [sr.to_dict() for sr in self.startup_rules]
        }

    def add_rule(self, rule: ActionRule) -> ActionRule:
        """Add a new action rule"""
        if not rule.id:
            rule.id = generate_rule_id('rule')
        self.rules.append(rule)
        return rule

    def remove_rule(self, rule_id: str) -> bool:
        """Remove a rule by ID"""
        for i, rule in enumerate(self.rules):
            if rule.id == rule_id:
                self.rules.pop(i)
                return True
        return False

    def update_rule(self, rule_id: str, updates: Dict[str, Any]) -> bool:
        """Update a rule by ID"""
        for rule in self.rules:
            if rule.id == rule_id:
                for key, value in updates.items():
                    if hasattr(rule, key):
                        setattr(rule, key, value)
                return True
        return False

    def add_group(self, group: ButtonGroup) -> ButtonGroup:
        """Add a new button group"""
        if not group.id:
            group.id = generate_rule_id('group')
        self.groups.append(group)
        return group

    def remove_group(self, group_id: str) -> bool:
        """Remove a group by ID"""
        for i, group in enumerate(self.groups):
            if group.id == group_id:
                self.groups.pop(i)
                return True
        return False

    def update_group(self, group_id: str, updates: Dict[str, Any]) -> bool:
        """Update a group by ID"""
        for group in self.groups:
            if group.id == group_id:
                for key, value in updates.items():
                    if hasattr(group, key):
                        setattr(group, key, value)
                return True
        return False

    def add_group_rule(self, rule: GroupRule) -> GroupRule:
        """Add a new group rule"""
        if not rule.id:
            rule.id = generate_rule_id('grouprule')
        self.group_rules.append(rule)
        return rule

    def remove_group_rule(self, rule_id: str) -> bool:
        """Remove a group rule by ID"""
        for i, rule in enumerate(self.group_rules):
            if rule.id == rule_id:
                self.group_rules.pop(i)
                return True
        return False

    def update_group_rule(self, rule_id: str, updates: Dict[str, Any]) -> bool:
        """Update a group rule by ID"""
        for rule in self.group_rules:
            if rule.id == rule_id:
                for key, value in updates.items():
                    if hasattr(rule, key):
                        setattr(rule, key, value)
                return True
        return False

    def get_group_rule_for_group(self, group_id: str) -> Optional[GroupRule]:
        """Get the group rule for a specific group (if any)"""
        for rule in self.group_rules:
            if rule.group_id == group_id:
                return rule
        return None

    def add_startup_rule(self, rule: StartupRule) -> StartupRule:
        """Add a startup rule"""
        if not rule.id:
            rule.id = generate_rule_id('startup')
        self.startup_rules.append(rule)
        return rule

    def remove_startup_rule(self, rule_id: str) -> bool:
        """Remove a startup rule by ID"""
        for i, rule in enumerate(self.startup_rules):
            if rule.id == rule_id:
                self.startup_rules.pop(i)
                return True
        return False

    def set_button_name(self, button_id: ButtonId, name: str) -> None:
        """Set a button's name"""
        if name:
            self.button_names[button_id] = name
        elif button_id in self.button_names:
            del self.button_names[button_id]

    def get_rules_for_button(self, button_id: ButtonId) -> List[ActionRule]:
        """Get all rules for a specific button"""
        return [r for r in self.rules if r.button == button_id]

    def get_group_for_button(self, button_id: ButtonId) -> Optional[ButtonGroup]:
        """Get the group that contains a specific button (if any)"""
        for group in self.groups:
            if button_id in group.buttons:
                return group
        return None

    def get_action_for_button_event(self, button_id: ButtonId, trigger: TriggerType) -> Optional[ActionDefinition]:
        """
        Get the action for a button press/release/hold event.
        Returns the action definition if found, or None if no matching rule.
        """
        # First check individual rules
        for rule in self.rules:
            rule_trigger = rule.trigger or 'release'
            if rule.button == button_id and rule_trigger == trigger:
                return rule.action
        
        # Then check groups - respect the trigger setting on the group rule (defaults to 'release')
        group = self.get_group_for_button(button_id)
        if group:
            try:
                button_index = group.buttons.index(button_id)
            except ValueError:
                button_index = -1
            
            if button_index >= 0:
                # Find the group rule for this group
                group_rule = self.get_group_rule_for_group(group.id)
                if group_rule:
                    # Check if the trigger matches (default to 'release' if not specified)
                    rule_trigger = group_rule.trigger or 'release'
                    if rule_trigger == trigger:
                        # Handle group action based on type
                        if group_rule.action.type == 'chord-progression':
                            # Return a set-chord-in-progression action
                            return ['set-chord-in-progression', group_rule.action.progression, button_index, group_rule.action.octave]
        
        return None


# Default empty configuration
DEFAULT_ACTION_RULES_CONFIG = ActionRulesConfig()
